File: Algorithm/FindServersThatHandledMostNumberofRequests.cpp
class Solution {
public:
    vector<int> busiestServers(int k, vector<int>& arrival, vector<int>& load) {
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> q; // {release_time, server}
        set<int> servers;
        vector<int> requests(k);
        
        for (int i = 0; i < k; ++i)
            servers.insert(i);
        
        for (int i = 0; i < arrival.size(); ++i) {
            const int t = arrival[i];
            const int l = load[i];
            
            // Release servers. O(logk) per pop()
            while (!q.empty() && q.top().first <= t) {
                servers.insert(q.top().second);
                q.pop();
            }
            
            // Drop the request.
            if (servers.empty()) continue;
            
            // Find first available one O(logk)
            auto it = servers.lower_bound(i % k);
            if (it == servers.end()) it = begin(servers);
            const int idx = *it;
            
            ++requests[idx];
            servers.erase(it);
            q.emplace(t + l, idx);
        }
        const int max_req = *max_element(begin(requests), end(requests));
        vector<int> ans;
        for (int i = 0; i < k; ++i)
            if (requests[i] == max_req) ans.push_back(i);
        return ans;
    }
};

File: Algorithm/FindServersThatHandledMostNumberofRequests.java
// Solution: Heap + TreeSet
// Use a min heap to store the release time -> server.
// Use a treeset to track the current available servers.
// For reach request, check whether servers can be released at that time.
// Time complexity: O(nlogk)
// Space complexity: O(k)
class Solution {
    public List<Integer> busiestServers(int k, int[] arrival, int[] load) {
        PriorityQueue<int[]> q = new PriorityQueue<int[]>((a, b) -> a[0] - b[0]); // {release_time, server}
        TreeSet<Integer> servers = new TreeSet<Integer>();
        int[] requests = new int[k];
        
        for (int i = 0; i < k; ++i)
            servers.add(i);
        
        for (int i = 0; i < arrival.length; ++i) {
            final int t = arrival[i];
            final int l = load[i];
            
            // Release servers. O(logk) per pop()
            while (!q.isEmpty() && q.peek()[0] <= t) {
                servers.add(q.peek()[1]);
                q.poll();
            }
            
            // Drop the request.
            if (servers.isEmpty()) continue;
            
            // Find first available one O(logk)
            Integer it = servers.ceiling(i % k); // 注意TreeSet和TreeMap的ceiling和ceilingKey或者ceilingEntry的API不同。
            if (it == null) it = servers.first();
            final int idx = it;
            
            ++requests[idx];
            servers.remove(it);
            q.offer(new int[]{t + l, idx});
        }
        int max_req = Integer.MIN_VALUE;
        for (int e : requests)
            max_req = Math.max(max_req, e);
        List<Integer> ans = new ArrayList<Integer>();
        for (int i = 0; i < k; ++i)
            if (requests[i] == max_req) ans.add(i);
        return ans;
    }
}
